lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -287,13 +287,20 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
                     $dqlAlias = $cacheKeyInfo['dqlAlias'];
                     $type     = $cacheKeyInfo['type'];
 
+                    if(
+                        isset($cacheKeyInfo['discriminatorColumn']) && 
+                        isset($data[$cacheKeyInfo['discriminatorColumn']]) &&
+                        $data[$cacheKeyInfo['discriminatorColumn']] != $cacheKeyInfo['discriminatorValue']
```

Comment:
@fullbl it shouldn't break it... let's keep the strict comparison and try to find the root cause.